src/internal/consistenthashing/etcd.go

```diff
@@ -0,0 +1,370 @@
+package consistenthashing
+
+import (
+	"context"
+	"fmt"
+	"hash/crc32"
+	"path"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/gogo/protobuf/types"
+	"github.com/pachyderm/pachyderm/v2/src/internal/collection"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/uuid"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/dlock"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/watch"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/sirupsen/logrus"
+	etcd "go.etcd.io/etcd/client/v3"
+)
+
+const (
+	// LeaseTTL is used by nodes in a ring to refresh their liveness state.
+	LeaseTTL = int64(5) // number of seconds between lease renewals.
+)
+
+// Ring is a consistent hash Ring. Each process should only create one instance with a given prefix. Nodes can be
+// added locally to a Ring. A Ring watches for changes to the prefix to determine if new nodes have been added
+// or deleted.
+// A Ring should be initialized with a call to New.
+type Ring struct {
+	Client *etcd.Client   // Client will be used to make api calls against etcd.
+	Prefix string         // Prefix is prepended to node keys stored in etcd to avoid naming conflicts.
+	Logger *logrus.Logger // Logger is used to output log messages if defined.
+
+	Cancel context.CancelFunc // Cancel can be used bya caller to stop Ring subroutines.
+	Ctx    context.Context    // Ctx is the context used to create the Ring.
+
+	node    node     // node represents the member instance in a Ring that is local to the current process.
+	members []Member // members is a string array of all node Ids in the Ring sorted by hashFn.
+
+	hashFn func([]byte) uint32 // hashFn will be called to determine the hash for a given key.
+
+}
+
+// node is a local Member of a Ring. Upon creation, a node requests a lease from etcd and continuously refreshes it
+// until it is shutdown.
+type node struct {
+	Member
+	Locks sync.Map // Locks is a thread-safe map of lockInfos associated with a node.
+}
+
+// Member contains metadata about members in a Ring.
+type Member struct {
+	Id   string // Id is the name of a members.
+	Hash uint32 // Hash is the output of the ring's hashFn(Id).
+}
+
+// lockInfo instances are created for keys by a ring's node.
+// A context lockInfo.Ctx is stored so the proper context is used when unlocking lockInfo.Lock
+type lockInfo struct {
+	Lock dlock.DLock     // Lock is a dlock that locks a key with a prefix in etcd.
+	Ctx  context.Context // Ctx is the context returned by Lock.
+}
+
+// New is the default constructor for Ring. Creating a Ring starts a goroutine to watch prefix in etcd and creates a
+// node that continuously refreshes a lease in etcd.
+func New(ctx context.Context, prefix string, client *etcd.Client, logger *logrus.Logger, id ...string) (*Ring, error) {
```

Comment:
This is more of a stylistic thing in the codebase, but I think it improves readability. We generally define parameters ordered by decreasing scope, so this would look more like `func New(ctx context.Context, client *etcd.Client, logger *logrus.Logger, prefix string, id ...string) (*Ring, error)`